Interpret strings as branches > semver constraints

carolynvs opened the pull request:
Fixes #710. When a user supplied string in an imported config file, or specified to `dep ensure`, can be interpreted multiple ways, prefer the branch over a semver constraint.

In #710, `glide.yaml` specified `v2` for `https://github.com/go-mgo/mgo`. When we assume that is a semver constraint, solve fails because the hinted revision in the lock (a commit on the v2 branch) doesn't satisfy the assumed constraint of `^2.0.0`.

The new preferred match order for the user string is:
* revision
* branch
* semver constraint
* tag

I am giving preference of a semver constraint over a tag so that a bare version, `1.0.0`, is interpreted more loosely with an implied caret, `^1.0.0`, instead of the stricter exact match.

carolynvs commented:
This isn't ready yet, doesn't work with the testdata from https://gist.github.com/ellotheth/e96f8f2c5fdd1b9e09e183970cfb276b.

carolynvs commented:
@sdboyer This is ready for review, though beware that it is currently pulling into two other PRs, #729 and #744. I recommend limiting the github diff to just the last commit so you don't see the changes from those other PRs.

I needed another test case for deduceConstraint where a branch name could be interpreted as a semver. I added that to my deptest fork but let me know if it belongs elsewhere.

carolynvs commented:
Rebased now that the other PRs are now merged.

carolynvs commented:
@sdboyer @darkowlzz @ibrasho Bumping for great glory. 🇺🇸 🎇

ibrasho commented on cmd/dep/root_analyzer.go:
```diff
@@ -150,19 +150,37 @@ func (a *rootAnalyzer) Info() (string, int) {
+// project based on the locked revision and the constraint from the manifest.
+// First try matching the revision to a version, then try the constraint from the
+// manifest, then finally the revision.
+func lookupVersionForLockedProject(pi gps.ProjectIdentifier, c gps.Constraint, rev gps.Revision, sm gps.SourceManager) (version gps.Version, warning error) {
```
What's the benefit of named return values here?

Also, naming the `error` `warning` seems a bit misleading.

For `lookupVersionForLockedProject`, it's an error that prevents returning the expected results. Consumers cannot rely on the version returned since it will be `nil`.

A consumer can call this value a `warning` if it doesn't depend on getting a valid `gps.Version` value. But I would only call it `warning` here if consumers could, somehow, use the returned value if they didn't care about the warning.

carolynvs replied:
> I would only call it warning here if consumers could, somehow, use the returned value if they didn't care about the warning.

Thanks! That is a good way of thinking about err/warn, very helpful.

ibrasho commented on cmd/dep/godep_importer.go:
```diff
@@ -125,11 +125,10 @@ func (g *godepImporter) convert(pr gps.ProjectRoot) (*dep.Manifest, *dep.Lock, e
-				warn := errors.Wrapf(err, "Unable to lookup the version represented by %s in %s. Falling back to locking the revision only.", pkg.Rev, pi.ProjectRoot)
-				g.logger.Printf(warn.Error())
-				version = revision
+			version, warning := lookupVersionForLockedProject(pi, nil, revision, g.sm)
```
I'd even call this variable `err` since the comment below explains why it was logged and ignored (same thing in glide importer).

carolynvs commented:
I've renamed from warning -> err and removed the named return values.

darkowlzz commented:
I went through it. LGTM 👍
Just wondering why were the files renames?
`cmd/dep/testdata/glide/glide.lock → cmd/dep/testdata/glide/glide0.lock`, etc

carolynvs commented:
> Just wondering why were the files renamed?

I added a second testdata file, e.g. glide1.lock, then didn't need it later on, but the filename change persisted after I squashed all my commits. Fixing.... 😊

carolynvs commented:
* @darkowlzz reminded me that[ I promised that this PR would also fix the godep importer](https://github.com/golang/dep/pull/742#discussion_r121284440).
* Improved which version is picked when a rev has multiple tags. When the manifest has a semver constraint, dep uses the constraint to filter the tagged versions. Since this only impacts the preferred version in the lock, the solver may still pick something different, but at least dep no longer suggests (and prints) something that is impossible. This stems from case in github.com/sdboyer/deptest, where rev `ff2948a` is tagged with both `v0.8.0` and `v1.0.0`.

darkowlzz commented on cmd/dep/root_analyzer_test.go:
```diff
@@ -19,3 +24,90 @@ func TestRootAnalyzer_Info(t *testing.T) {
+	wantV := "v1.0.0"
+	gotV := v.String()
+	if gotV != wantV {
+		t.Fatalf("Expected '%s', got '%s'", wantV, gotV)
```
How about adding a little context to the failure message, like we try to do everywhere and in alignment with the existing test in the file?
Same for the tests below 🙂

carolynvs commented:
@darkowlzz  Thanks for calling me out on that! 😀 I've added context to the test failure messages and squashed.

carolynvs commented:
Derp, my rebase went awry. All fixed (hopefully). 🤞

sdboyer commented:
This seems good to me - can merge once we deal with those conflicts 😄

carolynvs commented:
I've fixed the merge conflict.
